### PR TITLE
Don't emit CA2014 when stackalloc is in loop initializer

### LINKED ISSUE #6723: Message CA2014 is Misleading
I'm working on the Ecma V8 C# spec, and have the following test code:

```csharp
for (Span<int> sp1 = stackalloc int[2]; false;) { }
```

The use of `stackalloc` results in CA2014, whose explanation states, "Do not use `stackalloc` in loops." As my usage is part of the loop initializer, and is executed once only, it is **not** inside the loop. This can be demonstrated by rewriting my code as follows:

```csharp
Span<int> sp2 = stackalloc int[2];
for (; false;) { }
```

for which no message is issued.

I contend that the message should *not* be issued in my first case.

Rex

---
#### Document Details

⚠ *Do not edit this section. It is required for learn.microsoft.com ➟ GitHub issue linking.*

* ID: ea2c3912-46d6-54d9-61d2-6143853aa322
* Version Independent ID: 0e4e351e-f206-efcf-4d83-dd61bd9edd91
* Content: [CA2014: Do not use stackalloc in loops (code analysis) - .NET](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2014)
* Content Source: [docs/fundamentals/code-analysis/quality-rules/ca2014.md](https://github.com/dotnet/docs/blob/main/docs/fundamentals/code-analysis/quality-rules/ca2014.md)
* Product: **dotnet-fundamentals**
* GitHub Login: @stephentoub
* Microsoft Alias: **stoub**

### PR DESCRIPTION
**Pull request by CollinAlpert**
Fixes #6723

### CODE REVIEW

File: src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/CSharpDoNotUseStackallocInLoops.cs [hunk 1]
```diff
@@ -23,6 +23,12 @@ public sealed override void Initialize(AnalysisContext context)
             context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
             context.RegisterSyntaxNodeAction(ctx =>
             {
+                // stackalloc is in the for loop initializer and thus not part of the actual loop.
+                if (ctx.Node.Parent is EqualsValueClauseSyntax { Parent: VariableDeclaratorSyntax { Parent: VariableDeclarationSyntax { Parent: ForStatementSyntax } } })
+                {
+                    return;
+                }
+
```
Comment (stephentoub) [inline src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/CSharpDoNotUseStackallocInLoops.cs:31]: Won't this result in false negatives if this loop is nested in another?

I'd expect we'd instead want this at another special-case like https://github.com/dotnet/roslyn-analyzers/pull/6728/files#diff-b65a77da1c4644408673499e08231ec7bed47aa0a42e1b7097911f74ba19cfe1L38

Comment (CollinAlpert) [reply src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/CSharpDoNotUseStackallocInLoops.cs]: Yes, you're right. I've added it as a special case, just like `foreach` and added another test for it.

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DoNotUseStackallocInLoopsTests.cs [hunk 2]
```diff
@@ -306,5 +307,35 @@ public static void Foo()
...
+        [Fact, WorkItem(6723, "https://github.com/dotnet/roslyn-analyzers/issues/6723")]
+        public Task NoDiagnostics_StackallocInLoopInitializer()
+        {
+            return VerifyCS.VerifyAnalyzerAsync(@"
+                using System;
+                public class C
+                {
+                    public static void Foo()
+                    {
+                        for (Span<int> sp1 = stackalloc int[2]; false;) { }
```
Comment (stephentoub) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DoNotUseStackallocInLoopsTests.cs:320]: Can you also add tests that show:
- It still triggers when the stackalloc is in the condition or update expressions
- It doesn't trigger when a variable other than the first contains the stackalloc
- It doesn't trigger when the initializer contains the stackalloc as a subexpression, eg `Span<int> sp1 = Helper(stackalloc int[2])`

**Review by stephentoub — APPROVED** [review]
Thanks!

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DoNotUseStackallocInLoopsTests.cs (hunk 2 above)
Comment (CollinAlpert) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DoNotUseStackallocInLoopsTests.cs:320]: Sure! Done.

**Review by stephentoub — APPROVED** [review]
Thanks

**Pull request merged**